### PR TITLE
Added cgid capture to stirling with testing

### PR DESCRIPTION
**Pull request by AdityaAtulTewari**
Summary: Added Capturing of Cgroup ids to the ebpf probes in the socket tracer. This includes modifications to data structures for communicating that information to userspace and other data tables. It also includes some minimal testing for ensuring that cgroup ids are captured correctly. The reason for this change is given in the design document for the solution to this [issue](https://github.com/pixie-io/pixie/issues/1638): in short it is the first step in using cgids rather than pids to associate a particular piece of communication data with the correct pod.

Test Plan: Added Cgroup id capture checks to existing `socket_tracer` tests.

Type of change: /kind Feature expansion

Signed-off-as: AdityaAtulTewari [adityaatewari@gmail.com](mailto:adityaatewari@gmail.com)

### CODE REVIEW

**Comment by ddelnano** [comment]
@pixie-io/maintainers in the meantime would we be able to let the CI jobs run? @AdityaAtulTewari is interested in seeing if there are any failures to address.

File: src/shared/metadata/cgroup_path_resolver.cc
Comment (vihangm) [inline src/shared/metadata/cgroup_path_resolver.cc:98]: ditto on absl::Substitute usage here.

File: src/shared/metadata/cgroup_path_resolver.cc
Comment (AdityaAtulTewari) [reply src/shared/metadata/cgroup_path_resolver.cc]: I used StrCat it seemed more natural I hope that is appropriate.

File: src/shared/metadata/cgroup_path_resolver.cc
Comment (AdityaAtulTewari) [reply src/shared/metadata/cgroup_path_resolver.cc]: ditto for StrCat

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc
Comment (AdityaAtulTewari) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:1701]: Not populating the table causes a template compilation error iirc. Happy to discuss other failure mechanisms.

**Comment by AdityaAtulTewari** [comment]
> We had previously discussed writing a trace bpf test that would exhibit this short lived process behavior. I think it's important to develop that before we make the change to store the cgroup ids (this PR). Essentially my thoughts are that we would take your containerized application from [this branch](https://github.com/AdityaAtulTewari/pixie/tree/be0bcdd827116122da4f8f8ef6d6dd5fe2fac328/trigger_slp_error), create a new trace bpf test with it and then follow it up with this change to verify it results in the records we expect.

This change does not fix the race condition, it just exposes the cgroup id so that a future pr can. The changes are broken up this way as it allows me to break apart the solution to this problem into easier to design pieces. I'm happy to add that test to the change in progress that fixes the issue. However I don't think it is appropriate here because it will still fail.

File: src/stirling/source_connectors/socket_tracer/bcc_bpf_intf/common.h
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/bcc_bpf_intf/common.h:82]: I think we should remove this TODO. There are definitely valid reasons for making the cgid more of a useful construct for the rest of Pixie, but I think that specific solution is going to require a more detailed proposal and discussion.

File: src/stirling/source_connectors/socket_tracer/conn_stats.h
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/conn_stats.h:78]: Minor comment: My understanding is that this should never change. So it seems strange to label it as "current". I would prefer this to be just `cgid`.

**Review by ddelnano — COMMENTED** [review]
Two more additional comments on my latest round of review.

File: src/stirling/source_connectors/socket_tracer/conn_stats.h
Comment (AdityaAtulTewari) [reply src/stirling/source_connectors/socket_tracer/conn_stats.h:78]: This can change if the process migrates cgroups. While this is not possible explicitly in kubernetes nothing stops someone from making this behavior materialize in an actual linux program, within kubernetes.

Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/conn_stats.h:78]: Thanks for calling that out. I was vaguely aware of that, but didn't make the connection at first.

**Review by ddelnano — APPROVED** [review]
The test suite still needs to pass, but the changes look good to me pending that 👍

We will need to regenerate the data table docs with the following 
 tool: `//src/carnot/docstring/tables:datatables`. Happy to help with that after this is merged.

**Comment by ddelnano** [comment]
cc @etep since I believe you are interested in reviewing this as well.

**Review by etep — APPROVED** [review]

**Review by vihangm — APPROVED** [review]

**Pull request merged**
